### PR TITLE
Add MET and METplus

### PR DESCRIPTION
**Pull request by kgerheiser**
This draft PR replaces @jprestop's #169 and adds MET, METPLUS, and the GNU Scientific Library (GSL, a dependency) 

Most of it is straightforward, but MET has to link to Python through the variables `MET_PYTHON_CC` and `MET_PYTHON_LD`. These can be set manually in the config, but if they are not it attempts to use `python3-config --libs/--cflags` to fill them (maybe an option disable Python?).

`python3-config` can also be set through `MET_PYTHON_CONFIG` 

I have successfully tested it on Hera and Orion. Currently testing other systems and waiting for the CI to finish.

### CODE REVIEW

File: libs/build_gsl.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+set -eux
+
+name="gsl"
+version=${1:-${STACK_gsl_version}}
+
+compiler=$(echo $HPC_COMPILER | sed 's:/:-:g')
+mpi=$(echo $HPC_MPI | sed 's:/:-:g')
```
Comment (aerorahul) [inline libs/build_gsl.sh]: This is unnecessary.
```suggestion
```

File: libs/build_gsl.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+set -eux
+
+name="gsl"
+version=${1:-${STACK_gsl_version}}
+
+compiler=$(echo $HPC_COMPILER | sed 's:/:-:g')
```
Comment (aerorahul) [inline libs/build_gsl.sh:8]:
```suggestion
compiler=$(echo $HPC_COMPILER | sed 's/\//-/g')
```

File: libs/build_gsl.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    prefix="${PREFIX:-"/opt/modules"}/$compiler/$name/$version"
+    if [[ -d $prefix ]]; then
+      [[ $OVERWRITE =~ [yYtT] ]] && ( echo "WARNING: $prefix EXISTS: OVERWRITING!";$SUDO rm -rf $prefix ) \
+                                 || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+    fi
+else
+    prefix=${GSL_ROOT:-"/usr/local"}
+fi
+
+
```
Comment (aerorahul) [inline libs/build_gsl.sh]:
```suggestion
```

File: libs/build_gsl.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+../configure --prefix=$prefix
+
+make -j${NTHREADS:-4}
+[[ $MAKE_CHECK =~ [yYtT] ]] && make check
+$SUDO make install
+
+modpath=compiler
+$MODULES && update_modules $modpath $name $version \
+         || echo $name $version >> ${HPC_STACK_ROOT}/hpc-stack-contents.log
```
Comment (aerorahul) [inline libs/build_gsl.sh]:
```suggestion
         || echo $name $version $url >> ${HPC_STACK_ROOT}/hpc-stack-contents.log
```

File: libs/build_met.sh [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+LDFLAGS2="${MET_NETCDF}/lib:${MET_HDF5}/lib:${MET_BUFRLIB}"
+LDFLAGS3="${MET_GRIB2CLIB}:${MET_PYTHON}/lib:${MET_GSL}/lib"
+LDFLAGS4="-L${LIB_JASPER} -L${MET_HDF5}/lib -L${LIB_LIBPNG} -L${LIB_Z}"
+
+export LDFLAGS="${LDFLAGS2:-} ${LDFLAGS3:-} ${LDFLAGS4:-}"
+export LIBS="-lhdf5_hl -lhdf5 -lz"
+
+export CFLAGS+="-D__64BIT__"
+export CXXFLAGS+="-D__64BIT__"
```
Comment (aerorahul) [inline libs/build_met.sh]: Please move up where `CFLAGS`, `FFLAGS` are set.

File: libs/build_met.sh [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+cd ${HPC_STACK_ROOT}/${PKGDIR:-"pkg"}/${pkg_name}
+curr_dir=$(pwd)
+
+
+./configure --prefix=$prefix BUFRLIB_NAME=${BUFRLIB_NAME} GRIB2CLIB_NAME=${GRIB2CLIB_NAME} --enable-grib2 --enable-python
+
+make
+[[ $MAKE_CHECK =~ [yYtT] ]] && make check
+$SUDO make install
```
Comment (aerorahul) [inline libs/build_met.sh:119]: Is there no module file for `met`?

File: libs/build_met.sh [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+pkg_name=$name-$version
+url="https://github.com/dtcenter/MET/releases/download/v$version/$software.tar.gz"
+[[ -d $software ]] || ( $WGET $url; tar -xf $software.tar.gz )
+[[ ${DOWNLOAD_ONLY} =~ [yYtT] ]] && exit 0
+[[ -d $pkg_name ]] && cd $pkg_name || ( echo "$pkg_name does not exist, ABORT!"; exit 1 )
+
+cd ${HPC_STACK_ROOT}/${PKGDIR:-"pkg"}/${pkg_name}
+curr_dir=$(pwd)
+
+
```
Comment (aerorahul) [inline libs/build_met.sh]:
```suggestion
```

File: libs/build_met.sh (hunk 6 above)
Comment (aerorahul) [reply libs/build_met.sh:119]: Need something like:
```
# generate modulefile from template
[[ -z $mpi ]] && modpath=compiler || modpath=mpi
$MODULES && update_modules $modpath $name $version
echo $name $version $url >> ${HPC_STACK_ROOT}/hpc-stack-contents.log
```

File: libs/build_metplus.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+# Update the path to the MET tools for the users
+cd ${curr_dir}/parm/metplus_config
+cat metplus_system.conf | \
+  sed "s%MET_INSTALL_DIR = /path/to%MET_INSTALL_DIR = $met_prefix/met/$met_version%g" \
+  > metplus_system_new.conf
+mv metplus_system_new.conf metplus_system.conf
+
+
+
```
Comment (aerorahul) [inline libs/build_metplus.sh]: Where is the metplus modulefile creation?

File: libs/build_metplus.sh [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+pkg_name=$name-$version
+met_version=${1:-${STACK_metplus_version}}
+url="https://github.com/dtcenter/METplus/archive/$software.tar.gz"
+[[ -d $software ]] || ( $WGET $url; tar -xf $software.tar.gz )
+[[ -d $pkg_name ]] && cd $pkg_name || ( echo "$pkg_name does not exist, ABORT!"; exit 1 )
+
+
+# Hyphenated version used for install prefix
+compiler=$(echo $HPC_COMPILER | sed 's/\//-/g')
+mpi=$(echo $HPC_MPI | sed 's/\//-/g')
```
Comment (aerorahul) [inline libs/build_metplus.sh]: Why are these required?
They are not used to load or build anything here.

File: libs/build_metplus.sh [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+if $MODULES; then
+    prefix="${PREFIX:-"/opt/modules"}/$compiler/$name/$version"
+    mkdir -p $prefix
+    met_prefix="${PREFIX:-"/opt/modules"}/$compiler/$mpi"
+else
+    prefix=${MET_ROOT:-"/usr/local"}
+    met_prefix=$prefix
+fi
```
Comment (aerorahul) [inline libs/build_metplus.sh]: I don't understand this.
`prefix` is on the compiler hierarchy, while `met_prefix` is on mpi?

File: libs/build_metplus.sh [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+# Install is simply copying over the unpacked package to the install location
+cp -r ${HPC_STACK_ROOT}/${PKGDIR:-"pkg"}/${pkg_name}/* $prefix
+cd $prefix
+curr_dir=$(pwd)
+export PATH=${curr_dir}/ush:${PATH}
```
Comment (aerorahul) [inline libs/build_metplus.sh]: what is being used from `PATH` in this script from this point onward?

File: modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/met/met.lua [hunk 12]
```diff
@@ -0,0 +1,34 @@
+help([[
```
Comment (aerorahul) [inline modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/met/met.lua]: Need one in the `compiler` hierarchy as well, in case we build a serial version.

**Review by aerorahul — CHANGES_REQUESTED** [review]
See comments.

File: libs/build_gsl.sh (hunk 2 above)
Comment (kgerheiser) [reply libs/build_gsl.sh:8]: Isn't it a lot cleaner without escaping the forward slash?

Comment (aerorahul) [reply libs/build_gsl.sh:8]: Yes. It is a lot cleaner. 
I was going for consistency, but we can go forth with this from now on.

**Comment by jprestop** [comment]
@kgerheiser and @aerorahul I just wanted to follow up and see if I can do anything to help us make further progress on this or  should I just let @kgerheiser continue to make progress and answer questions when they arise?

**Comment by jprestop** [comment]
@kgerheiser and @aerorahul I just wanted to check in again and see if I can do anything to help us make further progress on this or if I should just let @kgerheiser continue to make progress and answer questions when they arise?

**Comment by kgerheiser** [comment]
Hey @jprestop, I've made progress to get the environment setup.

The issue I'm finding is that there is no portable way to link/include libpython/Python.h. I'm thinking it'll have to be manually set for each machine.

**Comment by jprestop** [comment]
> Hey @jprestop, I've made progress to get the environment setup.

Thank you so much!

> The issue I'm finding is that there is no portable way to link/include libpython/Python.h. I'm thinking it'll have to be manually set for each machine.

Yes, I suspected that might be the case and might be necessary.  It'd be great if "python3-config --ldflags" and "python3-config --cflags" always gave the correct information, but that is not always the case.

**Comment by jprestop** [comment]
Hi @kgerheiser.  Do you have an estimate on when you might be able to work on the Python configuration?  Do you need anything from me in order to make more progress?

File: libs/build_metplus.sh (hunk 9 above)
Comment (kgerheiser) [reply libs/build_metplus.sh]: Compiler is used for the install path

**Comment by WalterKolczynski-NOAA** [comment]
Where does this PR stand? We may have need of it soon for global-workflow (along with Issue #408).

**Comment by kgerheiser** [comment]
I've been working on this on and off over the past few weeks and it's more or less complete, and our upcoming Spack build includes MET/METplus.

**Comment by WalterKolczynski-NOAA** [comment]
Any movement on this? I've got global workflow using my own custom-built stack as a stopgap, but I'd still like to move to an official version.

**Comment by JessicaMeixner-NOAA** [comment]
@kgerheiser any chance this will be ready soon?  As @WalterKolczynski-NOAA mentioned, we're using his stop-gap in the workflow but with the recent updates in the version of intel in ufs-weather-model we cannot use that anymore.  This is likely going to be needed for the workflow PR: https://github.com/NOAA-EMC/global-workflow/pull/831 and is relevant for the workflow issue https://github.com/NOAA-EMC/global-workflow/issues/812

**Review by aerorahul — APPROVED** [review]
I think this is ready and looks good.
